File: backend/app/services/brief_outcomes.py
"""
Brief outcome calibration (Phase 3).

The closed loop: the web layer measures how a published brief actually performed
(actualScore / outcomeDelta vs the creator's baseline). This module reads that
history back so the brief-generation and idea-augmentation prompts can anchor
their predictions to what really happened for THIS creator — making scores
calibrated and visibly self-improving.
"""
import logging
import statistics
from typing import Optional

from app.db.pg_client import PgClient

logger = logging.getLogger(__name__)


def _median(values: list[float]) -> float:
    return float(statistics.median(values)) if values else 0.0


def fetch_creator_baseline(creator_id: str) -> float:
    """Median view count across the creator's videos — their performance baseline."""
    try:
        with PgClient() as db:
            db.execute(
                '''SELECT "viewCount" FROM "Vlog"
                   WHERE "creatorId" = %s AND "viewCount" IS NOT NULL AND "viewCount" > 0''',
                (creator_id,),
            )
            rows = db.fetchall() or []
        return _median([float(r["viewCount"]) for r in rows])
    except Exception as e:
        logger.warning("fetch_creator_baseline failed for %s: %s", creator_id, e)
        return 0.0


def fetch_calibration_context(creator_id: str) -> dict:
    """
    Return calibration context for prompts:
      - baseline_median_views: the creator's typical performance
      - samples: recent measured briefs (predicted vs actual)
      - mean_abs_error: how far past predictions were off, in score points
    Always returns a dict; empty/zeroed on any failure.
    """
    samples: list[dict] = []
    try:
        with PgClient() as db:
            db.execute(
                '''SELECT title, "estimatedScore", "actualScore", "outcomeDelta"
                   FROM "ContentBrief"
                   WHERE "creatorId" = %s AND "actualScore" IS NOT NULL
                   ORDER BY "measuredAt" DESC NULLS LAST
                   LIMIT 10''',
                (creator_id,),
            )
            samples = [dict(r) for r in (db.fetchall() or [])]
    except Exception as e:
        logger.warning("fetch_calibration_context failed for %s: %s", creator_id, e)

    baseline = fetch_creator_baseline(creator_id)

    mean_abs_error: Optional[float] = None
    if samples:
        errors = [
            abs((s.get("estimatedScore") or 0) - (s.get("actualScore") or 0))
            for s in samples
        ]
        if errors:
            mean_abs_error = sum(errors) / len(errors)

    return {
        "baseline_median_views": baseline,
        "samples": samples,
        "mean_abs_error": mean_abs_error,
    }


def format_calibration_section(calibration: Optional[dict]) -> str:
    """
    Render calibration context as a prompt section. Returns "" when there is no
    useful signal so prompts stay clean for brand-new creators.
    """
    if not calibration:
        return ""
    samples = calibration.get("samples") or []
    baseline = calibration.get("baseline_median_views") or 0
    if not samples and not baseline:
        return ""

    lines: list[str] = []
    if baseline:
        lines.append(f"Creator baseline: median {int(baseline):,} views per video.")
    for s in samples[:6]:
        delta = s.get("outcomeDelta")
        delta_note = (
            f"{delta * 100:+.0f}% vs baseline"
            if isinstance(delta, (int, float))
            else "outcome n/a"
        )
        lines.append(
            f"- Predicted {s.get('estimatedScore')}/100 for \"{s.get('title')}\" "
            f"→ actual {s.get('actualScore')}/100 ({delta_note})"
        )
    mae = calibration.get("mean_abs_error")
    if mae is not None:
        lines.append(
            f"Historical prediction error (mean absolute): {mae:.0f} points. "
            f"Calibrate new scores against these real outcomes — do not over-promise."
        )

    return (
        "\n\nCALIBRATION — THIS CREATOR'S ACTUAL OUTCOMES (predicted vs real):\n"
        + "\n".join(lines)
    )

File: apps/web/lib/briefOutcomes.ts
// Pure helpers for the brief outcome loop (Phase 3).
// When a content brief is linked to a published vlog, we measure how it actually
// performed against the creator's baseline so future predictions can be calibrated.
// No side effects — deterministic transformations only.

export function median(values: number[]): number {
  const nums = values
    .filter((v): v is number => typeof v === 'number' && !Number.isNaN(v))
    .sort((a, b) => a - b)
  if (nums.length === 0) return 0
  const mid = Math.floor(nums.length / 2)
  return nums.length % 2 ? nums[mid] : (nums[mid - 1] + nums[mid]) / 2
}

// Relative performance vs the creator's typical video: +0.5 == 50% above median.
// Returns null when there is no baseline to compare against.
export function computeOutcomeDelta(
  vlogViews: number,
  baselineMedian: number,
): number | null {
  if (!baselineMedian || baselineMedian <= 0) return null
  return (vlogViews - baselineMedian) / baselineMedian
}

// Map a relative delta onto a 0–100 score so it is comparable with estimatedScore.
// tanh keeps it bounded and smooth: delta 0 → 50, +1.0 (2× baseline) → ~88,
// -1.0 → ~12. Clamped to [0, 100].
export function outcomeToScore(delta: number | null): number | null {
  if (delta === null) return null
  const score = Math.round(50 + 50 * Math.tanh(delta))
  return Math.max(0, Math.min(100, score))
}

export type BriefOutcome = {
  actualScore: number | null
  outcomeDelta: number | null
}

// Measure a published vlog against the creator's other videos.
export function computeBriefOutcome(
  vlogViews: number,
  otherVlogViews: number[],
): BriefOutcome {
  const baseline = median(otherVlogViews)
  const outcomeDelta = computeOutcomeDelta(vlogViews, baseline)
  return { outcomeDelta, actualScore: outcomeToScore(outcomeDelta) }
}
